File: Tutorial1-1DaggerBasics/src/main/java/com/example/tutorial1daggerbasics/MainActivity.java
package com.example.tutorial1daggerbasics;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;
import android.widget.Toast;

import com.example.tutorial1daggerbasics.model.MyExample;

import java.util.Date;

import javax.inject.Inject;

public class MainActivity extends AppCompatActivity {

    // Note: This is a singleton object, MainActivity injects same object after device is rotated!
    // Note2: it's a singleton object if and only it's in a Application level module and annotated with @Singleton
    @Inject
    MyExample mMyExample;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        TextView dateTextView = findViewById(R.id.tv_date);

        ((MyApplication) getApplication())
                .getMyComponent()
                .inject(MainActivity.this);

        dateTextView.setText((new Date(mMyExample.getDate())).toString());

        System.out.println("MyExample: " + mMyExample);

        Toast.makeText(this, "MyExample: " + mMyExample, Toast.LENGTH_SHORT).show();
    }
}

File: Tutorial6-3SubComponentFactory/src/main/java/com/smarttoolfactory/tutorial6_3subcomponentfactory/MainActivity.kt
package com.smarttoolfactory.tutorial6_3subcomponentfactory

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.smarttoolfactory.tutorial6_3subcomponentfactory.model.DummyDependency
import javax.inject.Inject

/**
 * Component and SubComponent Factory can be used to replace Builders
 *
 * For Application Component create SubComponent factories with XSubComponent.Factory
 * and SubComponents should have inject() and create() methods to be able to inject to
 * Activity, Fragment or objects.
 *
 * 🔥 @ActivityScope on [ActivityScopedFragment] does not mean their objects live through
 * Activity lifecycle. Whenever [ActivityScopedFragment] fragment is replaced
 * new dependencies are created!!!
 */
/*
 ONLY one component can inject to an Object
 */
class MainActivity : AppCompatActivity() {

    // Injected from ApplicationModule
    @Inject
    lateinit var sharedPreferences: SharedPreferences

    // Injected from DummyDependencyModule
    @Inject
    lateinit var dummyDependency: DummyDependency


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initInjection()

        bindViews()


    }

    private fun bindViews() {
        findViewById<Button>(R.id.button_second_activity).setOnClickListener {
            val intent = Intent(this@MainActivity, SecondActivity::class.java)
            startActivity(intent)
        }

        findViewById<Button>(R.id.btn_first_frag).setOnClickListener {
            replaceFirstFragment()
        }
        replaceFirstFragment()

        findViewById<Button>(R.id.btn_second_frag).setOnClickListener {
            replaceSecondFragment()
        }
        replaceSecondFragment()

        findViewById<TextView>(R.id.tvInfo).text = "DummyDependenyc: ${dummyDependency.hashCode()}\n" +
                "sharedPreferences: ${sharedPreferences.hashCode()}"
        println("MainActivity sharedPreferences: $sharedPreferences")
        Toast.makeText(this, "MainActivity: $sharedPreferences", Toast.LENGTH_SHORT).show()
    }

    private fun replaceFirstFragment() {
        val ft = supportFragmentManager.beginTransaction()
        ft.replace(R.id.container1, FragScopedFragment()).commit()
    }

    private fun replaceSecondFragment() {
        val ft = supportFragmentManager.beginTransaction()
        ft.replace(R.id.container2, ActivityScopedFragment()).commit()
    }

    private fun initInjection() {

        val appComponent = (application as MyApplication).appComponent

        // 🔥SubComponent Builder
//        val dummyDependencyComponent = appComponent
//                .dummyDependencyBuilder()
//                .context(this)
//                .build()

        // 🔥 SubComponent Factory
        val dummyDependencyComponent = appComponent
                .dummyDependencyComponentFactory
                .create(this)

        dummyDependencyComponent.inject(this)
    }


}
